server/ecore-glsp/src/main/java/com/eclipsesource/glsp/ecore/operationhandler/EcoreLabelEditOperationHandler.java

```diff
@@ -87,20 +92,45 @@ public void execute(AbstractOperationAction action, GraphicalModelState graphica
 					((EEnumLiteral) eObject).setName(inputText);
 				}
 			}
-		} else { // Main Label of a Node
-			GNode node = getOrThrow(index.findElementByClass(editLabelAction.getLabelId(), GNode.class),
-					"No label container for label with id " + editLabelAction.getLabelId() + " found");
-
-			eObject = getOrThrow(index.getSemantic(node),
+		} else { 
+			Optional<GNode> nodeOptinal = index.findElementByClass(editLabelAction.getLabelId(), GNode.class);
+			if (nodeOptinal.isPresent()) { // Main Label of a Node
+				GNode node = nodeOptinal.get();
+				eObject = getOrThrow(index.getSemantic(node),
 					"No semantic element for labelContainer with id " + node.getId() + " found");
 
-			Shape shape = getOrThrow(index.getNotation(eObject), Shape.class,
-					"No shape element for label with id " + editLabelAction.getLabelId() + " found");
+				Shape shape = getOrThrow(index.getNotation(eObject), Shape.class,
+						"No shape element for label with id " + editLabelAction.getLabelId() + " found");
+
+				if (eObject instanceof EClassifier) {
+					((EClassifier) eObject).setName(editLabelAction.getText().trim());
+					// nameChange== uri change so we have to recreate the proxy here
+					shape.setSemanticElement(facade.createProxy(eObject));
+				}
 
-			if (eObject instanceof EClassifier) {
-				((EClassifier) eObject).setName(editLabelAction.getText().trim());
-				// nameChange== uri change so we have to recreate the proxy here
-				shape.setSemanticElement(facade.createProxy(eObject));
+			} else { // Reference
```

Comment:
isn't there a better way to identify whether we are editing a reference or not?

Reply:
At this point, we know we're editing the label of an edge, because we excluded all other possible cases (The label itself is not a semantic element, e.g. an EAttribute/EOperation, and it's not the label of a Node, e.g. a Class Name/Interface Name; so it can only be the label of an Edge). But testing if it's indeed the label of an EReference is tested a few lines below.

Still, I would prefer a more cautious approach than "getOrThrow", and simply get an Optional<GEdge> and test if it's present; as we never really know which labeled elements may be added in the future (To make this crash, we'd need to edit a Label that is not a semantic element, and is owned neither by a Node nor an Edge. This is unlikely to happen in the future, but not totally impossible)

Reply:
@CamilleLetavernier We can use the more cautious approch, I only reused what was done before with the label of a node.

@eneufeld  I'm also not 100% satisfied how it is identified but as we have to check if a parent is a GEdge I did not find a better solution. This is the only other way I can think of:
`
Optional<?> optional;
if ((optional = index.findElementByClass(editLabelAction.getLabelId(), GNode.class)).isPresent()) {
    GNode node = (GNode) optional.get();
    // ...
} else if ((optional = index.findElementByClass(editLabelAction.getLabelId(), GEdge.class)).isPresent()) {
    GEdge edge = (GEdge) optional.get();
    // ...
}
`

Reply:
I agree that we should guard against a label that is neither a child of a node or an edge.

I like the proposal of @leodennis but maybe it would even be more clear if we'd switch over the label type (`label:edge` vs. `label:text`).

`type = index.findElementByClass(editLabelAction.getLabelId(), GModelElement.class).map(e -> e.getType())`

If `type` is absent, we can't continue anyway, because that means that there is no element with that ID.
After that, I think we are good to rely on the type string and throw a runtime exception if something is missing, because that means we ran into a bug anyway (i.e., the graph factory generated a wrong graph).

Even better, we might want to assign even more concrete label types, such as `label:reference`, `label:class-name`, and `label:attribute`. This way, it is pretty clear which case we are handling.

Of course, we'd have to make sure the graph factory uses those labels accordingly and we'd have to register those types in https://github.com/eclipsesource/ecore-glsp/blob/master/client/sprotty-ecore/src/di.config.ts so that they are rendered properly.